crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+            ast::Expr::Tuple(ast::ExprTuple { elts, range, .. }) => (elts, range),
+            _ => return None,
+        };
+        // An `__all__` definition with <2 elements can't be unsorted;
+        // no point in proceeding any further here
+        if elts.len() < 2 {
+            return None;
+        }
+        for elt in elts {
+            // Only consider sorting it if __all__ only has strings in it
+            let string_literal = elt.as_string_literal_expr()?;
+            // And if any strings are implicitly concatenated, don't bother
+            if string_literal.value.is_implicit_concatenated() {
+                return None;
+            }
+        }
+        // Step (2): parse the `__all__` definition using the raw tokens.
+        //
+        // (2a). Start by collecting information on each line individually:
+        let lines = collect_dunder_all_lines(*range, locator)?;
+
+        // (2b). Group lines together into sortable "items":
+        //   - Any "item" contains a single element of the `__all__` list/tuple
+        //   - "Items" are ordered according to the element they contain
+        //   - Assume that any comments on their own line are meant to be grouped
+        //     with the element immediately below them: if the element moves,
+        //     the comments above the element move with it.
+        //   - The same goes for any comments on the same line as an element:
+        //     if the element moves, the comment moves with it.
+        let items = collect_dunder_all_items(&lines);
+
+        Some(DunderAllValue {
+            items,
+            range: *range,
+            multiline: is_multiline,
+        })
+    }
+
+    fn construct_sorted_all(
+        &self,
+        locator: &Locator,
+        parent: &ast::Stmt,
+        stylist: &Stylist,
+    ) -> SortedDunderAll {
+        let mut sorted_items = self.items.clone();
+        sorted_items.sort();
+
+        // As well as saving us unnecessary work,
+        // returning early here also means that we can rely on the invariant
+        // throughout the rest of this function that both `items` and `sorted_items`
+        // have length of at least two. If there are fewer than two items in `__all__`,
+        // it is impossible for them *not* to compare equal here:
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
+        assert!(self.items.len() >= 2);
+
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis
+        //    (if there was one), up to the start of the first element in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
```

Comment:
Nit: One thing I like to do is to add a python snippet to give readers a concrete example:

```suggestion
        // As well as the "items" in the `__all__` definition,
        // there is also a "prelude" and a "postlude":
        //  - Prelude == the region of source code from the opening parenthesis
        //    (if there was one), up to the start of the first element in `__all__`.
        //  - Postlude == the region of source code from the end of the last
        //    element in `__all__` up to and including the closing parenthesis
        //    (if there was one).
        // 
        // ```python
        // __all__ = [
        //   "first item",
        //   "last item"   
        // ]
        // ```
        // The prelude in the above example is the text from right after the `[` up to `"first item"`. The postlude is the text between `"last item"` and `]`.
        let prelude_end = {
```

Reply:
I added a variant of this in 5e129ade8e77090a2aa1017611258b2c2f65d942. (By the way: writing that up was what made me spot the bug regarding multiline comments that I fixed in fd1dc54c80004d1b2377ca93ad30718d83e77587!)